pkg/security/ptracer/process.go

```diff
@@ -0,0 +1,101 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+
+// Package ptracer holds the start command of CWS injector
+package ptracer
+
+import "github.com/DataDog/datadog-agent/pkg/security/proto/ebpfless"
+
+// Resources defines shared process resources
+type Resources struct {
+	Fd  map[int32]string
+	Cwd string
+}
+
+// Process represents a process context
+type Process struct {
+	Pid  int
+	Tgid int
+	Nr   map[int]*ebpfless.SyscallMsg
+	Res  *Resources
+}
+
+// NewProcess returns a new process
+func NewProcess(pid int) *Process {
+	return &Process{
+		Pid:  pid,
+		Tgid: pid,
+		Nr:   make(map[int]*ebpfless.SyscallMsg),
+		Res: &Resources{
+			Fd: make(map[int32]string),
+		},
+	}
+}
+
+// ProcessCache defines a thread cache
+type ProcessCache struct {
+	pid2Process map[int]*Process
+	tgid2Pid    map[int]map[int]bool
+}
+
+// NewProcessCache returns a new thread cache
+func NewProcessCache() *ProcessCache {
+	return &ProcessCache{
+		pid2Process: make(map[int]*Process),
+		tgid2Pid:    make(map[int]map[int]bool),
+	}
+}
+
+// Add a process
+func (tc *ProcessCache) Add(pid int, process *Process) {
+	tc.pid2Process[pid] = process
+
+	pids, ok := tc.tgid2Pid[process.Tgid]
+	if !ok {
+		pids = make(map[int]bool)
+		tc.tgid2Pid[process.Tgid] = pids
+	}
+
+	pids[pid] = true
+}
+
+// SetAsThreadOf set the process as thread of the given tgid
+func (tc *ProcessCache) SetAsThreadOf(process *Process, tgid int) {
+	if parent := tc.pid2Process[tgid]; parent != nil {
+		// share resources, parent should never be nil
+		process.Res = parent.Res
+	}
+	process.Tgid = tgid
+
+	// re-add to update the caches
+	tc.Add(process.Pid, process)
+}
+
+// Remove a pid
+func (tc *ProcessCache) Remove(pid int) {
+	delete(tc.pid2Process, pid)
+
+	pids, ok := tc.tgid2Pid[pid]
+	if !ok {
+		return
+	}
+	delete(tc.tgid2Pid, pid)
+
+	for pid = range pids {
+		tc.Remove(pid)
+	}
+}
+
+// Get return the process entry for the given pid
+func (tc *ProcessCache) Get(pid int) *Process {
+	return tc.pid2Process[pid]
+}
+
+// Tgid returns the tgid of the given pid
+func (tc *ProcessCache) Tgid(pid int) int {
+	return tc.pid2Process[pid].Tgid
```

Comment:
The Get() dont need to check the value accessed by the map, as it can return nil.
But this one could segfault if we don't check that there is an entry for the given pid (even if could never happen)

Reply:
Going to remove this one, not used